### PR TITLE
add support for fastify

### LINKED ISSUE #287: Module plugin request: Fastify
https://www.fastify.io

It would be nice to have this supported. While it is compatible with Express middleware in general, and similar implementation as Express could probably work... Fastify does offer its own [request / response hooks](https://www.fastify.io/docs/latest/Hooks/) that may be more appropriate to utilize at the top level scope.

### PR DESCRIPTION
**Pull request by rochdev**
### What does this PR do?

Adds support for Fastify

### Motivation

Close #287 

### Plugin Checklist

- [x] Unit tests.
- [x] TypeScript [definitions][1].
- [x] TypeScript [tests][2].
- [x] API [documentation][3].
- [x] CircleCI [jobs/workflows][4].
- [x] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/src/plugins/index.js

### CODE REVIEW

File: packages/datadog-plugin-fastify/src/index.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  return function wrapFastify (fastify) {
+    const fastifyWithTrace = function fastifyWithTrace () {
+      return fastify._datadog_wrapper.apply(this, arguments)
+    }
+
+    fastify._datadog_wrapper = function () {
+      const app = fastify.apply(this, arguments)
+
+      app.addHook('onRequest', createOnRequest(tracer, config))
+      app.addHook('preHandler', preHandler)
```
Comment (rishabh) [inline packages/datadog-plugin-fastify/src/index.js]: Why do we hook into the `preHandler` and not the `onSend` or `onError`.

Also, can we also wrap `send` using the `reply` in `onRequest`?

File: packages/datadog-plugin-fastify/test/index.spec.js [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        after(() => {
+          return agent.close()
+        })
+
+        beforeEach(() => {
+          fastify = require(`../../../versions/fastify@${version}`).get()
+        })
+
+        it('should do automatic instrumentation on the app routes', done => {
+          app = fastify()
```
Comment (rishabh) [inline packages/datadog-plugin-fastify/test/index.spec.js]: Can we move `app = fastify()` to a `beforeEach`?

File: packages/datadog-plugin-fastify/src/index.js (hunk 1 above)
Comment (rochdev) [reply packages/datadog-plugin-fastify/src/index.js]: > Why do we hook into the preHandler and not the onSend or onError.

This is to make sure we wrap before the handler, and `preHandler` is the hook that happens before the handler and receives a Fastify request in the most consistent way. `onError` doesn't exist in older versions, and `onSend` happens to late in the lifecycle.

> Also, can we also wrap send using the reply in onRequest?

In older versions, `onRequest` didn't pass `request` and instead passed the raw request.

File: packages/datadog-plugin-fastify/test/index.spec.js (hunk 2 above)
Comment (rochdev) [reply packages/datadog-plugin-fastify/test/index.spec.js]: Done.

File: packages/datadog-plugin-fastify/src/index.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    patch (fastify, tracer, config) {
+      // `fastify` is a function so we return a wrapper that will replace its export.
+      return createWrapFastify(tracer, config)(fastify)
+    },
+    unpatch (fastify) {
+      unwrapFastify(fastify)
+    }
+  },
+  {
+    name: 'find-my-way',
```
[the remark this answers is not in the document]
Comment (rochdev) [reply packages/datadog-plugin-fastify/src/index.js]: Done.

**Pull request merged**